### ISSUE TITLE
variablesResolutionMode: 20210326 failing to resolve variable with multiple defaults

### ISSUE THREAD

**Issue by perrin4869** [serverless/serverless#9880]
I have been working on upgrading my project from v1 to v2, and the variable resolution broke on a certain condition.
When there are multiple defaults, if the second one fails to resolve, the whole resolution will fail.

<details>
<summary><code>serverless.yml</code></summary>

```yaml
service: aws-node-project

frameworkVersion: '2'
variablesResolutionMode: 20210326

provider:
  name: aws
  runtime: nodejs12.x
  stage: ${opt:stage, env:STAGE, env:AWS_STAGE, "local"}
  lambdaHashingVersion: 20201221

functions:
  hello:
    handler: handler.hello

```

</details>

<details>
<summary><b><code>npx serverless package</code> output</b></summary>

```
❯ npx serverless package

 Serverless Error ----------------------------------------

  Cannot resolve serverless.yml: "provider.stage" property is not accessible (configured behind variables which cannot be resolved at this stage)

  Get Support --------------------------------------------
     Docs:          docs.serverless.com
     Bugs:          github.com/serverless/serverless/issues
     Issues:        forum.serverless.com

  Your Environment Information ---------------------------
     Operating System:          linux
     Node Version:              14.17.1
     Framework Version:         2.54.0 (local)
     Plugin Version:            5.4.3
     SDK Version:               4.2.6
     Components Version:        3.15.0

```

</details>

Running `STAGE=foo npx serverless package` will work correctly since then the first default will resolve.

<b>Installed version</b>

```
Framework Core: 2.54.0 (local)
Plugin: 5.4.3
SDK: 4.2.6
Components: 3.15.0
```

**Comment by medikoo** [comment]
@perrin4869 thanks for report. 

You're receiving that error because `stage` plays part in resolution of environment variables, e.g. you may rely on `.env.<stage>` file which contains environment variables configuration and to load variables from there, first we need to resolve the stage.

We've also discussed very same problem here: https://github.com/serverless/serverless/issues/9813

Ideally if `stage` value is not controlled in anyway by environment variables. What is the use case for doing that on your side?

**Comment by perrin4869** [comment]
Well, I use environmental variables in my CI, which are set differently based on the pipeline. The staging pipeline will have `STAGE=staging` and the production pipeline will have `STAGE=production`. I do need to reevaluate the usage of `AWS_STAGING` variable.

**Comment by medikoo** [comment]
> The staging pipeline will have STAGE=staging and the production pipeline will have STAGE=production. I do need to reevaluate the usage of AWS_STAGING variable.

Do you rely on one CI configuration for different pipelines? If such you may also solve it by keeping setting the `STAGE`, but by propagating it to `sls` via `--stage=$STAGE` flag.

Note that `provider.stage` setting in a configuratioin is technically a setting for _default stage_. It's a last resort fallback if there's no `--stage` flag passed (the default is normally `dev`, but if you wish to change it, it's when `provider.stage` should be configured). We also have proposal of renaming this to `provider.defaultStage` -> https://github.com/serverless/serverless/issues/8684 so it's not confusing, but we're also considering totally removing support for this property (as it's hard to find a good use case for `dev` not being a default stage)

Additionally if you want to rely on _resolved_ stage in configuration you should simply refer to `sls:stage` (until it was introduced developers quite often ensured that actual stage was resolved on `provider.stage`, and then referring to that in variables, but it was just dirty workaround for fact that there wasn't anything like `sls:stage` available. It is now

**Comment by perrin4869** [comment]
I see! Yeah I've been using `provider.stage` since like forever, didn't notice `sls:stage` until recently :)
Indeed changing it to `provider.defaultStage` would clarify the intended use of the option.
Still, there are practical difficulties not being able to use the env `STAGE` as a default for stage.

Usually I do not run `serverless` directly, but rather define a script inside `package.json`, like `"deploy": "serverless deploy --verbose"`. This would mean changing the scripts to be rather `"deploy": "serverless deploy --verbose --stage $STAGE"`. The difficulty here is, what if I run this locally, and I do not have $STAGE defined? In this case I want to default to `"local"` or `"development"` or something, which is what I was achieving by doing `stage: ${opt:stage, env:STAGE, "local"}`.

Maybe `stage` could be a special case, where it does rely on an environment variable when setting it the first time?

**Comment by medikoo** [comment]
> Maybe stage could be a special case, where it does rely on an environment variable when setting it the first time?

Problem is that we do not want to introduce a confusing contradiction. Currently this change prevents such scenario from happening. 

Let's imagine following service setup:

_serverless.yml_
```yaml
provider:
  stage: ${env:FOO, 'dev'}
  otherImportantSetting: ${env:FOO}
...
```

_.env.dev_
```sh
FOO=bar
```

And let's imagine running it without `FOO` environment variable set, and without `--stage` flag

Current flow is that we need to resolve `stage` first and as no `--stage` flag is present, we need to resort to _default_ which is customized at `provider.stage`, and this depends on eventually `FOO` to be set.

Now if in this specific case we'll assume that `FOO` is not set, it eventually still can be set via `.env<stage>` file.

And in this case we will end with resolved configuration as:


```yaml
provider:
  stage: dev
  otherImportantSetting: bar
```

Which logically should not happen.

So this validation is to prevent such edge confusing results to surface.

--- 

However, as I think now about it more deeply, technically we may try to improve handling of this by allowing to ignore missing env setting eventually needed for `provider.stage` resolution, and throw only if we find that assumed as not set environment variables is happened to be actually configured in eventually loaded `.env*` file.

So in other words instead of unconditionally blocking such construct to prevent possible contradiction, we may throw only when we detect that contradicting configuration has been placed, and I think it's possible to solve it technically with not big effort

@pgrzesik what do you think?

**Comment by perrin4869** [comment]
I can totally appreciate how tough this is :)
Yeah, that's the example as to why `stage` would be a special case, so it is not an ideal situation ><;;
I'll be happy whichever way you guys choose!

**Comment by perrin4869** [comment]
@medikoo thanks!! This seems perfect!
Looking forward to giving this a try :)

**Comment by lkolchin** [comment]
Hi,

Can someone please tell me why the heck is this not in the doco?
```
Do you rely on one CI configuration for different pipelines? If such you may also solve it by keeping setting the STAGE, but by propagating it to sls via --stage=$STAGE flag.

Note that provider.stage setting in a configuratioin is technically a setting for default stage. It's a last resort fallback if there's no --stage flag passed (the default is normally dev, but if you wish to change it, it's when provider.stage should be configured). We also have proposal of renaming this to provider.defaultStage -> https://github.com/serverless/serverless/issues/8684 so it's not confusing, but we're also considering totally removing support for this property (as it's hard to find a good use case for dev not being a default stage)

Additionally if you want to rely on resolved stage in configuration you should simply refer to sls:stage (until it was introduced, developers quite often ensured that actual stage was resolved on provider.stage, and then referring to that in variables, but it was just dirty workaround for fact that there wasn't anything like sls:stage available. It is now)
```

I was using `${self:provider.stage}` forever but with SLS v3 it causes - `Cannot resolve serverless.yml: "provider.stage" property is not accessible (configured behind variables which cannot be resolved at this stage)` and really I had to Google extensively to find what the problem is because error description quite cryptic :(

So, replacing `${sls:stage}` helped.

Another issue is - looking through the doco:
This - https://www.serverless.com/framework/docs/providers/aws/guide/variables tells me that `${opt:stage, 'dev'}` is resolvable

But this - https://www.serverless.com/framework/docs/guides/upgrading-v3 tells me that `${opt:foo}` variable must be replaced with `${param:foo}`

So, is `--stage` a special case and is the only one that can be resolved with `${opt:stage}` or not, because this still works for me:
```
provider:
  stage: ${opt:stage, env:STAGE_NAME, 'lab' }
```
 In the rest of doco it still uses `${opt:region, self:provider.region, "us-east-1"}.` - so what's the deal with `${opt:}` and SLS v3?????
